### ISSUE TITLE
Parallel workers

### ISSUE THREAD

**Issue by cyno** [s3tools/s3cmd#2]
Starting point could be this nice fork by pcorliss who added parallel workers:
https://github.com/pcorliss/s3cmd-modification
In my case, this speeds up my upload by 50x times.
My use case is uploading 20.000.000 small files around several S3 locations (Pearltrees.com assets).

I also forked pcorliss modifications to add parallel workers to cp and mv commands:
https://github.com/cyno/s3cmd-modification

As you are currently working on the 1.1 release, even if the code changed from the fork, it would be very nice to resolve conflicts and merge this great fork.

**Comment by dieend** [comment]
Are we still doesn't have this feature? Downloading 20000 file tooks very long time because of no parallel capability.

**Comment by overthink** [comment]
+1

(s4cmd https://github.com/bloomreach/s4cmd is often MUCH faster due to concurrent workers... but it lacks other features that s3cmd has...)

**Comment by castro1688** [comment]
Please.. ??

**Comment by fubarhouse** [comment]
+1

**Comment by albarki** [comment]
+1 
parallel workers is an urgent, saves a lot of time for us.

**Comment by jlevy** [comment]
For those looking here: Although other tools don't exactly replicate everything from s3cmd, the best solution for anyone handling lots of files at once is probably to use [s4cmd](https://github.com/bloomreach/s4cmd) or [aws-cli](https://github.com/aws/aws-cli) (discussed [here](https://aws.amazon.com/blogs/apn/getting-the-most-out-of-the-amazon-s3-cli/)), both of which have parallel workers.

**Comment by dannyman** [comment]
Thanks, @jlevy ... do you know if aws-cli now supports large file transfers? The reason I needed to switch to s3cmd was to chunk large files.

**Comment by jlevy** [comment]
@dannyman yes, see the article linked in my last comment; it handles multipart.

Slight digression, but would be interested to see a benchmark of s4cmd vs aws-cli for common scenarios like in that blog — anyone want to run those or similar tests with both? 😃 There is some uncertainty around perf for all these tools and a good benchmark would help. Cc @chouhanyang and https://github.com/bloomreach/s4cmd/issues/72

**Comment by dannyman** [comment]
Thanks, @jlevy ... do you know if aws-cli now supports large file transfers? The reason I needed to switch to s3cmd was to chunk large files.

**Comment by isegal** [comment]
👍 parallel uploads are extremely useful. In our case using a highest tier C4, we were able to push thousands of files per second up to s3 using https://github.com/mishudark/s3-parallel-put so it would be nice to see it supported in s3cmd.

**Comment by tgmedia-nz** [comment]
same here - is there any progress?? :D

**Comment by mdomsch** [comment]
aws-cli does parallel uploads quite well.  We're unlikely to get that added
into s3cmd any time soon - it would be a fairly invasive upgrade, and the
limited volunteers making commits to s3cmd lately don't have the time for
that major overhaul.
